### PR TITLE
[Perf] MiniCPM-o 4.5 thinker cuda graph

### PR DESCRIPTION
**Pull request by nagisa-kunhah**
## Purpose

Part of #5069.

Enable `FULL_DECODE_ONLY` CUDA Graph for the MiniCPM-o 4.5 Stage 0 Thinker at batch size 1.

The profile keeps Stage 1 Talker and Stage 2 Code2Wav in eager mode. It also sets the FlashAttention CUDA Graph split count to 1 to preserve greedy decoding parity with the no-Graph path.

## Test Plan

**vLLM Version:** `0.26.0`

**vLLM-Omni Commit:** `d2e843ba19300f08a9877fb2648cf5c2cd1246aa`

**Hardware:** `2 x NVIDIA H20 97,871 MiB`

**Model:** `OpenBMB/MiniCPM-o-4_5`, BF16, FlashAttention

Performance benchmark:

- Random dataset
- Prefix length: 16 tokens
- Input length: 128 tokens
- Output length: 512 tokens, EOS ignored
- Requests: 10
- Concurrency: 1
- Sampling: `temperature=0`, `seed=42`
- Metrics: TTFT, TPOT, ITL, E2EL, and output throughput
- Graph capture size: 1
- Graph warmups: 2

Compared three Stage 0 configurations while keeping all other pipeline settings unchanged:

1. Compile only, CUDA Graph disabled:
   - `enforce_eager: false`
   - `cudagraph_mode: NONE`
2. FULL CUDA Graph with the default FlashAttention split count:
   - `cudagraph_mode: FULL_DECODE_ONLY`
   - `flash_attn_max_num_splits_for_cuda_graph: 32`
3. FULL CUDA Graph with the parity fix:
   - `cudagraph_mode: FULL_DECODE_ONLY`
   - `flash_attn_max_num_splits_for_cuda_graph: 1`

Accuracy was checked with a fixed 128-token greedy generation. Top-2 logits, selected token IDs, and final output hashes were compared against the compile-only/no-Graph baseline.

### Performance

All benchmark requests completed successfully.

| Stage 0 mode | FA splits | Mean TTFT | Mean TPOT | Mean ITL | Mean E2EL | Output throughput |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Compile only, no Graph | Dynamic | 53.46 ms | 17.57 ms | 17.57 ms | 9033.38 ms | 56.68 tok/s |
| FULL Graph | 32 | 42.50 ms | 7.40 ms | 7.40 ms | 3824.56 ms | 133.86 tok/s |
| FULL Graph | 1 | 43.05 ms | 8.12 ms | 8.12 ms | 4190.66 ms | 122.17 tok/s |

FULL Graph with 32 splits versus compile-only/no-Graph:

- TPOT reduction: `57.9%`
- Decode speedup: `2.37x`
- Output throughput increase: `136.2%`

FULL Graph with 1 split versus compile-only/no-Graph:

- TPOT reduction: `53.8%`
- Decode speedup: `2.16x`
- Output throughput increase: `115.6%`

Using 1 split is about `9.7%` slower in TPOT than 32 splits, while retaining most of the CUDA Graph speedup.

### Accuracy

| Configuration | Result versus compile-only/no-Graph |
| --- | --- |
| FULL Graph, 32 splits | Top-2 logits first differ at decode step 111. The selected token first differs at step 117. Final output hash differs. |
| FULL Graph, 1 split | Top-2 logits and selected token IDs match for all 128 decode steps. Final text and output hash match exactly. |

The 32-split difference comes from BF16 reduction-order changes in FlashAttention. FULL CUDA Graph requires a fixed split count, while the no-Graph path uses the dynamic split heuristic. The small numerical difference accumulates during autoregressive decoding and can flip a low-margin token.

Setting `flash_attn_max_num_splits_for_cuda_graph: 1` restores exact greedy parity for the validated batch-size-1 workload.

Focused tests:

```text
5 passed, 22 deselected
```

Ruff and `git diff --check` also pass.

**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.

(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by amy-why-3459** [comment]
## Summary

This PR enables `FULL_DECODE_ONLY` CUDA Graph for the MiniCPM-o 4.5 Stage 0 Thinker via a new opt-in deploy profile (`minicpmo_4_5_cudagraph.yaml`), and adds runner-level handling so the CUDA Graph replay tuple (a flattened NamedTuple) survives inter-stage extraction. The approach is well-scoped: the graph is gated behind an opt-in profile rather than flipping defaults, and the `flash_attn_max_num_splits_for_cuda_graph=1` parity work is validated carefully (2.37x decode speedup; byte-identical greedy output for all 128 decode steps with 1 split).

### Are the shared-module changes necessary?

A quick clarification on "shared/public" code: `vllm_omni/worker/gpu_model_runner.py` is **not** upstream vLLM code — it is vLLM-Omni's own `OmniGPUModelRunner` (subclass of vLLM's `GPUModelRunner`). So this PR does not touch upstream vLLM.

However, `extract_multimodal_outputs` is a **shared worker-layer method used by all Omni models**, called from 6 runners:

- `gpu_model_runner.py:1165`
- `gpu_ar_model_runner.py:1349`
- `gpu_generation_model_runner.py:367, 863`
- `npu_model_runner.py:319`
- `npu_ar_model_runner.py:726`
- `npu_generation_model_runner.py:401`

These runners serve qwen3-omni, moss-tts, voxtral, voxcpm2, audex, and other Omni models, so any change here has an all-model blast radius.

Importantly, the codebase already provides a **model-side hook** — `make_omni_output` — invoked in `_model_forward`:

```python
if not isinstance(model_output, OmniOutput) and hasattr(self.model, "make_omni_output"):
    model_output = self.model.make_omni_output(model_output, **model_kwargs, **model_kwargs_extra)
```

MiniCPM-o's `forward` already returns `OmniOutput`. So the "CUDA Graph flattens NamedTuple to a plain tuple" problem could be fixed **inside the model** (e.g. teach `make_omni_output` or `forward` to normalize the replay tuple back to `OmniOutput`), avoiding any change to the shared runner and narrowing the blast radius to MiniCPM-o only.

That said, since vLLM's CUDA graph output weak-reference mechanism flattens NamedTuples generally, a shared-runner fix is a reasonable non-breaking enhancement for any model returning a NamedTuple under FULL graphs. **My recommendation: either fix in the model hook (minimal blast radius) or, if keeping the runner branch, tighten its matching so it does not rely on fragile length/type field inference.**

---

## Requested changes

### 1. The new `extract_multimodal_outputs` tuple branch relies on fragile field inference (shared runner, affects all Omni models)

In `vllm_omni/worker/gpu_model_runner.py`, the new branch matches:

```python
elif (
    hasattr(self.model, "have_multimodal_outputs")
    and self.model.have_multimodal_outputs
    and isinstance(hidden_states, tuple)
    and len(hidden_states) == len(OmniOutput._fields)   # == 4
    and isinstance(hidden_states[0], torch.Tensor)
    and (hidden_states[1] is None or isinstance(hidden_states[1], dict))
):
```

Problems:

- Matching on "length == 4 and second element is a dict" does **not** guarantee the elements are in `OmniOutput` field order. If the weak-reference replay tuple is `(tensor, dict, tensor, tensor)` — with `intermediate_tensors`/`next_token_id` also tensors — the second element happens to be a dict and would be misread as `multimodal_outputs`, even though the positional semantics differ.
- This branch runs **before** the pre-existing generic list/tuple branch, so any model that emits a 4-tuple with a dict second element will be silently re-routed.

**Suggested fixes (pick one):**

- **Preferred:** fix at the model layer. MiniCPM-o already has the `make_omni_output` hook and already returns `OmniOutput` from `forward`. Normalize the replay tuple inside the model, leaving the shared runner untouched (blast radius narrowed to MiniCPM-o).
- If keeping the runner branch, confirm the actual tuple layout produced by vLLM's CUDA graph output weak-reference rebuild (see `cudagraph_outputs_weak_refs` handling of NamedTuples) rather than guessing by length; or add an explicit marker field to `OmniOutput` to identify a rebuilt tuple.

### 2. Thinker NPU branch uses a different `positions` condition than the CUDA branch

In `vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni.py`:

```python
if current_omni_platform.is_npu():
    thinker_positions = positions[0] if positions.ndim > 1 else positions
# CUDA path:
thinker_positions = positions[0] if positions is not None and positions_added_batch_dim else positions
```

The whole point of splitting `added_batch_dim` into three flags is to handle batched 3D M-RoPE positions `(batch, seq, 3)` correctly. The NPU branch still keys off `ndim > 1`, which mis-slices a batched 3D positions tensor `(1, seq, 3)` (where `positions_added_batch_dim is False`). Please unify on `positions_added_batch_dim` for the NPU path too, and add a test with `(1, seq, 3)` positions (the current test only covers `(3, 1)`).

### 3. New deploy profile constraints should be explicit

`vllm_omni/deploy/minicpmo_4_5_cudagraph.yaml` sets Stage 0 `max_num_seqs: 1` (Stages 1 and 2 also 1), unlike the default profile (4/4/4). Please:

- Add a header comment stating this is an **experimental validation profile** and that concurrency is bounded by the bs=1 graph capture size.
- Align `limit_mm_per_prompt.video` with the default profile (default uses `{count: 1, num_frames: 32}`; the new profile uses `video: 1`).
- Register the new profile in `_DEPLOY_LAYOUTS` in `tests/model_executor/models/minicpmo_4_5/test_pipeline.py` (it is currently missing).

---

## Suggestions (non-blocking)

### 4. Add a smoke test that actually captures/replays a CUDA graph

`test_cudagraph_forward.py` verifies that `forward` preserves tensor identity (necessary but not sufficient for graph capture) but never executes a real capture/replay. Consider adding a GPU CI smoke test (capture, replay once, compare outputs), or at least note in the PR description that this is a CPU-level unit test and real capture relies on the manual benchmark in the PR body.

### 5. Document the new profile

Mention the opt-in profile and its constraints in `recipes/OpenBMB/MiniCPM-o-4_5.md`, in the same style as the `token2wav_trt` guidance.

### 6. State the validation boundary explicitly

CI is green and the greedy-parity validation in the PR body is solid (1 split → exact parity over 128 steps). However, parity at `max_num_seqs > 1` is not validated — that is a known boundary of the bs=1-locked profile. Please state this limitation explicitly in the PR description.

File: tests/worker/test_omni_gpu_model_runner.py [hunk 1]
```diff
@@ -136,6 +137,26 @@ def test_filter_mrope_kwargs_preserves_flexible_model_kwargs():
...
+    text_hidden_states = torch.randn(2, 4)
+    latent = torch.randn(2, 4)
+    replay_output = tuple(
+        OmniOutput(
+            text_hidden_states=text_hidden_states,
+            multimodal_outputs={"latent": latent},
+        )
+    )
+
+    extracted_hidden_states, multimodal_outputs = runner.extract_multimodal_outputs(replay_output)
```
Comment (lishunyang12) [inline tests/worker/test_omni_gpu_model_runner.py:153]: Run the replay tuple through make_omni_output before testing extraction.
